text/0039-_formatted-field-behavior.md

```diff
@@ -0,0 +1,474 @@
+        }
+    ]
+}
+```
+
+Unlike v0.20, if no field are matched to be formatted, `_formatted` is not computed nor returned in response.
+
+**Example 3**
+
+Given these search parameters:
+
+```
+{
+    "q": "Prince",
+    "attributesToRetrieve": ["title"],
+    "attributesToHighlight": ["actor"]
+}
+```
+
+As a user I get:
+
+```
+{
+    "hits": [
+        {
+            "title": "Prince Avalanche",
+            "_formatted": {
+                "actor": "<em>Prince</em>"
+            }
+        }
+    ]
+}
+```
+
+`_formatted` is only filled with the fields set in `attributesToHighlight` despite the fact that the user only ask for `title` in `attributesToRetrieve`.
+
+Same behavior as v0.20 release.
+
+**Example 4**
+
+Given these search paramters:
+
+```
+{
+    "q": "Prince",
+    "attributesToRetrieve": ["actor", "title"],
+    "attributesToHighlight": ["actor"]
+}
+```
+
+As a user i get:
+
+```
+{
+    "hits": [
+        {
+            "title": "Prince Avalanche",
+            "actor": "Prince",
+            "_formatted": {
+                "actor": "<em>Prince</em>"
+            }
+        }
+    ]
+}
+```
+
+Same behavior as v0.20 release.
+
+**Example 5**
+
+Given these search parameters:
+```
+{
+    "q": "prince",
+    "attributesToRetrieve": ["*"],
+    "attributesToHighlight": ["title"]
+}
+```
+
+As a user i get:
+
+```
+{
+    "hits": [
+        {
+            "title": "Prince Avalanche",
+            "actor": "Prince",
+            "poster": "https://image.tmdb.org/t/p/w1280/3KHiQt54usbHyIjLIMzaDAoIJNK.jpg",
+            "_formatted": {
+                "title": "<em>Prince</em> Avalanche"
+            }
+        }
+    ]
+}
+```
+
+Unlike v0.20, `_formatted` is only containing fields set in `attributesToHighlight` and `attributesToCrop`. Independently from `attributesToRetrieve` value.
+
+**Example 6**
+
+Given these search parameters:
+```
+{
+    "q": "prince",
+    "attributesToRetrieve": ["title"],
+    "attributesToHighlight": ["*"]
+}
+```
+
+As a user i get:
+
+```
+{
+    "hits": [
+        {
+            "title": "Prince Avalanche",
+            "_formatted": {
+                "title": "<em>Prince</em> Avalanche",
+                "actor": "<em>Prince</em>,
+                "poster": "https://image.tmdb.org/t/p/w1280/3KHiQt54usbHyIjLIMzaDAoIJNK.jpg"
+            }
+        }
+    ]
+}
+```
+
+Unlike v0.20, `attributesToHighlight` set fields to be in `_formatted` independently from `attributesToRetrieve`.
+
+#### Expected MeiliSearch Behavior (0.21)
+
+✅ If `attributesToRetrieve` is not set as a parameter, the expected behavior is the same as if `attributesToRetrieve` is equal to `*`.
+✅ If `attributesToHighlight` and `attributesToCrop` are not set, do not return `_formatted` and don't compute highlights and crops.
+✅ If cumulated fields in `attributesToHighlight` and `attributesToCrop` resolve to only having non-existent fields, do not return `_formatted`.
```

Comment:
Could we send back an empty `_formatted`?  To avoid key checking? Not sure this is the place, maybe it should be determined for a later version. 

```
{
  "title": "hello",
  "_formatted": {}
}
```

Reply:
Thanks, @bidoubiwa! Can you explain a bit more about avoiding key checking? Btw, your proposal can be added to this specification iteration. :)

Reply:
@gmourier can you ping me if there is any change like this, I will open an issue once this spec is definitely validated

Reply:
I'm coding this part right now and this specific behavior would make the code more complex than it's already. Since the `_formatted` is currently a non-easy to maintain part in the code, I would rather we stay with "no `_formatted` if the fields are not valid".
If I found out a better solution to implement this, I'll come back to you and will update this spec :)

Reply:
Hello @curquiza! I'm sorry! I misspoke above when I said, "this can be added to the spec". I should have said that the specification was not merged and could still evolve! I was waiting for more information about the need expressed by @bidoubiwa about "avoiding key checking" before changing anything in the speicifcation! Anyway, thanks for the information 🙂

Reply:
Well, I cannot remember the reason why. So lets forget about it 🔥